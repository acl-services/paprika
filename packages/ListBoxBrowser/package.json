{
  "name": "@paprika/list-box-browser",
  "version": "1.0.11",
  "description": "ListBox Browser component is a two column layout that allows users to select one or multiple options. It is similar to the experience of browsing a file manager in any operating system.",
  "author": "@paprika",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/acl-services/paprika.git",
    "directory": "packages/ListBoxBrowser"
  },
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@babel/runtime-corejs2": "^7.3.1",
    "@paprika/helpers": "^2.1.0",
    "@paprika/icon": "^1.0.4",
    "@paprika/list-box": "^1.3.4",
    "@paprika/raw-button": "^1.0.3",
    "@paprika/spinner": "^1.0.5",
    "@paprika/stylers": "^0.1.5",
    "@paprika/tokens": "^1.0.0",
    "prop-types": "^15.7.2"
  },
  "peerDependencies": {
    "@paprika/l10n": "^1.1.3",
    "react": "^16.8.4",
    "react-dom": "^16.8.4",
    "styled-components": "^4.2.0"
  },
  "paprikaDocs": {
    "subComponents": [
      "CustomListBox",
      "Title"
    ]
  },
  "gitHead": "8bff2e8a6f01708ba60d9178f867dba378806370"
}
