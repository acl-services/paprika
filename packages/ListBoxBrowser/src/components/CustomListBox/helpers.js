export function isSelected({ $$key, selectedOptions, browserKey, isRootListBox }) {
  let key = browserKey;
  if (isRootListBox) {
    key = "root";
  }

  if (key in selectedOptions) {
    return selectedOptions[key].some(option => option.$$key === $$key);
  }

  return false;
}

export function isSelectable({ hasOptions, isParentSelectable }) {
  if (isParentSelectable !== null) {
    return !isParentSelectable;
  }

  return hasOptions;
}
