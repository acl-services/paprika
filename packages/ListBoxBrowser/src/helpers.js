import React from "react";

/** helpers */
const hasOptions = options => typeof options !== "undefined" && Array.isArray(options);

export function isRoot(key) {
  if (key === null) return false;

  return key.includes("root");
}

function getOptions(indexes, list) {
  if (Array.isArray(indexes)) {
    return indexes.map(index => list[index].value);
  }

  const options = [list[indexes].value].filter(chunk => chunk);
  return options.length ? options : null;
}

function isSelectable({ hasOptions, isParentSelectable }) {
  if (isParentSelectable !== null) {
    return !isParentSelectable;
  }

  return hasOptions;
}

/** DATA STRUCTURE */

export function getData({
  data,
  path = "root",
  refDefaultSelectedOptions,
  defaultSelectedOptions,
  defaultSelectedView,
  refDefaultSelectedView,
}) {
  return data.map(({ label = null, options, ...moreAttributes }, index) => {
    if (!label) {
      throw new Error("A 'label' attribute is required for each option object, can't process data.");
    }

    let newPath = isRoot(path) ? `${index}` : `${path}/${index}`;
    const recursiveParameters = {
      data: options,
      defaultSelectedOptions,
      defaultSelectedView,
      path: newPath,
      refDefaultSelectedOptions,
      refDefaultSelectedView,
    };
    const _hasOptions = hasOptions(options);

    const option = {
      parent: path,
      hasOptions: _hasOptions,
      $$key: newPath,
      attributes: {
        label,
        options: _hasOptions ? getData(recursiveParameters) : null,
        ...moreAttributes,
      },
    };

    if (defaultSelectedOptions(option.attributes)) {
      /* eslint-disable no-param-reassign */
      if (isRoot(option.parent)) {
        newPath = "root";
      }

      if (option.parent in refDefaultSelectedOptions.current) {
        refDefaultSelectedOptions.current = {
          ...refDefaultSelectedOptions.current,
          [option.parent]: [...refDefaultSelectedOptions.current[option.parent], option],
        };
      } else {
        refDefaultSelectedOptions.current = { ...refDefaultSelectedOptions.current, [option.parent]: [option] };
      }
    }

    // checking if its initial view
    if (_hasOptions && defaultSelectedView && defaultSelectedView(option.attributes)) {
      refDefaultSelectedView.current = newPath;
    }

    return option;
  });
}

export function getOptionByKey(data, path) {
  let node = null;
  function runner(data, path) {
    return data.map(option => {
      if (!option.$$key.includes("/") && typeof data[path] !== "undefined" && option.$$key === path) {
        node = option;
        throw new Error("option found");
      }

      if (option.$$key === path) {
        node = option;
        throw new Error("option found");
      }

      if (hasOptions(option.attributes.options)) {
        return runner(option.attributes.options, path);
      }

      return [];
    });
  }

  try {
    runner(data, path);
    return [];
  } catch (e) {
    return node;
  }
}

export function getBreadcrumb({ data, option }) {
  let activeOption = option;
  let key = null;
  const breadcrumb = [];
  while (!isRoot(key)) {
    if (!isRoot(activeOption.parent)) {
      const parent = getOptionByKey(data, activeOption.parent);
      breadcrumb.push(parent);
      activeOption = parent;
    }
    key = activeOption.parent;
  }

  return breadcrumb.reverse();
}

export function onChange({ source, indexes, list, isParentSelectable, setSelectedOptions, browserKey, isMulti }) {
  if (!isMulti) {
    setSelectedOptions(selectedOptions => {
      const key = isRoot(source) ? "root" : browserKey;
      if (Object.keys(selectedOptions).length) {
        const option = getOptions(indexes, list);
        if (option === null) return {};

        const isSameOption = Object.keys(selectedOptions).some(
          key => selectedOptions[key][0].$$key === option[0].$$key
        );

        if (isSameOption) {
          return {};
        }
      }

      const options = getOptions(indexes, list);
      if (options === null) return {};
      return { [key]: options };
    });

    return;
  }

  if (isRoot(source)) {
    const options = getOptions(indexes, list)
      .map(option => {
        if (isSelectable({ hasOptions: option.hasOptions, isParentSelectable })) {
          return null;
        }

        return option;
      })
      .filter(chunk => chunk);

    if (options.length) {
      setSelectedOptions(selectedOptions => ({ ...selectedOptions, root: [...options] }));
      return;
    }

    setSelectedOptions(selectedOptions => ({ ...selectedOptions, root: [] }));
  }

  setSelectedOptions(selectedOptions => ({ ...selectedOptions, [browserKey]: getOptions(indexes, list) }));
}

export function focusListBoxBrowser($root, hasLeftColumn) {
  window.requestAnimationFrame(() => {
    const index = hasLeftColumn ? 1 : 0;
    $root.querySelectorAll('[data-pka-anchor="list-box-content-inline"]')[index].focus();
  });
}

export function focusListBoxRoot($root) {
  window.requestAnimationFrame(() => {
    $root.querySelectorAll('[data-pka-anchor="list-box-content-inline"]')[0].focus();
  });
}

export function getInitialView(localData) {
  let index = null;
  try {
    Object.keys(localData).some(key => {
      if (
        localData[key].hasOptions ||
        (localData[key].attributes.options && localData[key].attributes.options.length === 0)
      ) {
        index = key;
        throw new Error("index found");
      }

      return false;
    });

    return index;
  } catch (e) {
    return index;
  }
}

export function extractedExtendedProps(children) {
  const props = {};
  React.Children.map(children, child => {
    if (child === null) return null;

    if (child.type.displayName === "ListBoxBrowser.Root") {
      props.root = { ...child.props };
    }

    if (child.type.displayName === "ListBoxBrowser.Browser") {
      props.browser = { ...child.props };
    }
  });

  return { ...props, children };
}

export function getDataOptionByFn(data, fn) {
  let node = null;
  function runner(data, fn) {
    data.forEach(option => {
      if (fn(option)) {
        node = option;
        throw new Error("We have found the option");
      }

      if (hasOptions(option.options)) {
        runner(option.options, fn);
      }
    });
  }

  try {
    runner(data, fn);
    return node;
  } catch (e) {
    // will fire if an option is found
    return node;
  }
}
