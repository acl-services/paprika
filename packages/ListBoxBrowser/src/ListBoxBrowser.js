import React from "react";
import PropTypes from "prop-types";
import { container, flex } from "./ListBoxBrowser.styles";
import {
  getData,
  getOptionByKey,
  onChange as onChangeHelper,
  isRoot,
  focusListBoxBrowser,
  focusListBoxRoot,
} from "./helpers";
import OptionsSelected from "./components/OptionsSelected";
import CustomListBox from "./components/CustomListBox";
import Title from "./components/Title";

const propTypes = {
  data: PropTypes.any.isRequired, // eslint-disable-line
  isMulti: PropTypes.bool,
  height: PropTypes.number,
  onChange: PropTypes.func,
  isParentSelectable: PropTypes.bool,
  rootTitle: PropTypes.string,
  browserTitle: PropTypes.string,
};

const defaultProps = {
  isMulti: true,
  height: 220,
  onChange: () => {},
  isParentSelectable: null,
  rootTitle: "",
  browserTitle: "",
};

export default function ListBoxBrowser(props) {
  const refSelectedOptions = React.useRef({});
  const { onChange, isParentSelectable, data, isMulti, height, rootTitle, browserTitle } = props;
  const [localData] = React.useState(() => getData({ data, selectedOptions: refSelectedOptions }));
  const [rootKey, setRootKey] = React.useState("0");
  // NOTE: will yield a bug if the first option has not children
  const [browserKey, setBrowserKey] = React.useState("0");
  const [selectedOptions, setSelectedOptions] = React.useState(refSelectedOptions.current);

  const browserOptions = React.useMemo(() => getOptionByKey(localData, browserKey || rootKey), [
    browserKey,
    localData,
    rootKey,
  ]);

  const handleChange = source => (indexes, list) => {
    onChangeHelper({
      source,
      indexes,
      list,
      isParentSelectable,
      setSelectedOptions,
      browserKey,
      isMulti,
    });
  };

  const handleClickRoot = ({ $$key, hasOptions, isClickFromButton = false }) => event => {
    if ((hasOptions && isParentSelectable === null) || isClickFromButton) {
      setRootKey($$key);
      setBrowserKey($$key);
      focusListBoxBrowser();
      if (event) event.stopPropagation();
    }
  };

  const handleClickBrowser = ({ $$key, isClickFromButton = false }) => event => {
    const option = getOptionByKey(localData, $$key);
    if ((option.hasOptions && isParentSelectable === null) || isClickFromButton) {
      setBrowserKey($$key);
      focusListBoxBrowser();
      if (event) event.stopPropagation();
    }
  };

  const handleUp = _parent => () => {
    setBrowserKey(`${_parent}`);
    focusListBoxBrowser();
  };

  function handleClickBreadcrumb(option) {
    setBrowserKey(`${option.$$key}`);
    focusListBoxBrowser();
  }

  function handleClickJumpToOption(option) {
    if (option.hasOptions) {
      const attributes = { $$key: option.$$key, hasOptions: true, isClickFromButton: true };
      if (isRoot(option.$$key)) {
        handleClickRoot(...attributes)();
        return;
      }
      handleClickBrowser({ ...attributes })();
      return;
    }

    if (isRoot(option.parent)) {
      focusListBoxRoot();
      return;
    }

    handleClickBrowser({ $$key: option.parent, hasOptions: false, isClickFromButton: true })();
  }

  function handleRemove(option) {
    setSelectedOptions(selectedOptions => {
      console.log(option);
      const cloneSelectedOptions = { ...selectedOptions };
      let index = null;
      cloneSelectedOptions[option.parent].some((opt, i) => {
        if (opt.$$key === option.$$key) {
          index = i;
          return true;
        }
        return false;
      });

      cloneSelectedOptions[option.parent].splice(index, 1);
      return cloneSelectedOptions;
    });
  }

  React.useEffect(() => {
    onChange(selectedOptions);
  }, [onChange, selectedOptions]);

  return (
    <div isParentSelectable={isParentSelectable} css={container} height={height}>
      <Title
        rootTitle={rootTitle}
        browserTitle={browserTitle}
        onClickBreadcrumb={handleClickBreadcrumb}
        browserKey={browserKey}
        data={localData}
      />
      <div css={flex}>
        <CustomListBox
          browserKey="root"
          height={height}
          isMulti={isMulti}
          onChange={handleChange("root")}
          onClickNavigate={handleClickRoot}
          options={localData}
          rootKey={rootKey}
          selectedOptions={selectedOptions}
          isParentSelectable={isParentSelectable}
        />
        <CustomListBox
          id={browserKey}
          browserKey={browserKey}
          hasOnUp={!isRoot(browserOptions.parent)}
          height={height}
          isMulti={isMulti}
          onChange={handleChange("browser")}
          onClickNavigate={handleClickBrowser}
          onUp={handleUp(browserOptions.parent)}
          options={browserOptions.attributes.options}
          selectedOptions={selectedOptions}
          isParentSelectable={isParentSelectable}
        />
      </div>

      <OptionsSelected
        onRemove={handleRemove}
        onClick={handleClickJumpToOption}
        data={localData}
        options={selectedOptions}
      />
    </div>
  );
}

ListBoxBrowser.propTypes = propTypes;
ListBoxBrowser.defaultProps = defaultProps;
