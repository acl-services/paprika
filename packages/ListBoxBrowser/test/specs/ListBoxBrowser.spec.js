import { render } from "@testing-library/react";
import React from "react";
import data from "./fixtures/multiple";
import dataMultipleFirstOptionNoOptions from "./fixtures/multiple.firstOptionNoOptions";
import dataNoOptions from "./fixtures/multiple.withNoOptions";
import { getData, getOptionByKey, getDataOptionByFn } from "../../src/helpers";
import ListBoxBrowser from "../../src";

const selectedOptions = { current: {} };
const dataParameters = { data, selectedOptions, defaultSelectedOptions: () => false };

describe("ListBoxBrowser", () => {
  it("should return a data with proper keys", () => {
    const modifiedData = getData(dataParameters);

    expect(modifiedData).toMatchSnapshot();
  });

  it("should return option with $$key === 0", () => {
    const option = getOptionByKey(getData(dataParameters), "0");
    expect(option).toMatchSnapshot();
  });

  it("should return option with $$key === 1", () => {
    const option = getOptionByKey(getData(dataParameters), "1");
    expect(option).toMatchSnapshot();
  });

  it("should return option with $$key === 2/1", () => {
    const option = getOptionByKey(getData(dataParameters), "2/1");
    expect(option).toMatchSnapshot();
  });

  it("should return option with $$key === 1/2/1", () => {
    const option = getOptionByKey(getData(dataParameters), "1/2/1");
    expect(option).toMatchSnapshot();
  });

  it("should render the ListBoxBrowser with selected option", () => {
    const { getByText } = render(
      <ListBoxBrowser data={dataMultipleFirstOptionNoOptions} rootTitle="Universes" browserTitle="Heroes">
        <ListBoxBrowser.OptionsSelected />
      </ListBoxBrowser>
    );

    const selectedOption = document.querySelector("[data-ppk-is-root-selected='true']");
    expect(selectedOption).toBeInTheDocument();
    expect(selectedOption.textContent).toBe("Starcraft Universe");
    expect(getByText(/Terrans/)).toBeInTheDocument();
    expect(getByText(/Zergs/)).toBeInTheDocument();
    expect(getByText(/Protos/)).toBeInTheDocument();
  });

  it("should throw Error with message 'At least one option should have options attribute' when none option has options", () => {
    expect.assertions(1);
    const error = console.error;
    console.error = () => {};
    try {
      render(
        <ListBoxBrowser data={dataNoOptions}>
          <ListBoxBrowser.OptionsSelected />
        </ListBoxBrowser>
      );
    } catch (e) {
      expect(e.message).toBe("At least one option should have options attribute");
      console.error = error;
    }
  });

  it("should find the options using getDataOptionByFn function", () => {
    const option = getDataOptionByFn(data, option => option.label === "Angelo Fortunato");

    const test1 = getDataOptionByFn(data, option => option.key === "test1");

    const test2 = getDataOptionByFn(data, option => option.key === "test2");

    expect(option).toMatchSnapshot();
    expect(test1).toMatchSnapshot();
    expect(test2).toMatchSnapshot();
  });
});
