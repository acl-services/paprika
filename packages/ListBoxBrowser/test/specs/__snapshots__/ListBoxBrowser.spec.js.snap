// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListBoxBrowser should find the options using getDataOptionByFn function 1`] = `
Object {
  "label": "Angelo Fortunato",
}
`;

exports[`ListBoxBrowser should find the options using getDataOptionByFn function 2`] = `
Object {
  "key": "test1",
  "label": "Mutants",
}
`;

exports[`ListBoxBrowser should find the options using getDataOptionByFn function 3`] = `
Object {
  "key": "test2",
  "label": "Celestials",
}
`;

exports[`ListBoxBrowser should return a data with proper keys 1`] = `
Array [
  Object {
    "$$key": "0",
    "attributes": Object {
      "label": "Marvel Universe",
      "options": Array [
        Object {
          "$$key": "0/0",
          "attributes": Object {
            "label": "Humans",
            "options": Array [
              Object {
                "$$key": "0/0/0",
                "attributes": Object {
                  "label": "Black Cat",
                  "options": Array [
                    Object {
                      "$$key": "0/0/0/0",
                      "attributes": Object {
                        "label": "Akhenaten",
                        "options": null,
                      },
                      "hasOptions": false,
                      "parent": "0/0/0",
                    },
                    Object {
                      "$$key": "0/0/0/1",
                      "attributes": Object {
                        "label": "Alistair Smythe",
                        "options": null,
                      },
                      "hasOptions": false,
                      "parent": "0/0/0",
                    },
                    Object {
                      "$$key": "0/0/0/2",
                      "attributes": Object {
                        "label": "Angelo Fortunato",
                        "options": null,
                      },
                      "hasOptions": false,
                      "parent": "0/0/0",
                    },
                    Object {
                      "$$key": "0/0/0/3",
                      "attributes": Object {
                        "label": "Arcade",
                        "options": null,
                      },
                      "hasOptions": false,
                      "parent": "0/0/0",
                    },
                    Object {
                      "$$key": "0/0/0/4",
                      "attributes": Object {
                        "label": "Aria",
                        "options": null,
                      },
                      "hasOptions": false,
                      "parent": "0/0/0",
                    },
                    Object {
                      "$$key": "0/0/0/5",
                      "attributes": Object {
                        "label": "Baal",
                        "options": null,
                      },
                      "hasOptions": false,
                      "parent": "0/0/0",
                    },
                  ],
                },
                "hasOptions": true,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/1",
                "attributes": Object {
                  "label": "Silver Sable",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/2",
                "attributes": Object {
                  "label": "Red Skull",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/3",
                "attributes": Object {
                  "label": "Vulture",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/4",
                "attributes": Object {
                  "label": "Falcon",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/5",
                "attributes": Object {
                  "label": "Kraven the hunter",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/6",
                "attributes": Object {
                  "label": "Bruce banner",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/7",
                "attributes": Object {
                  "label": "Moon Knight",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/8",
                "attributes": Object {
                  "label": "KingPin",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/9",
                "attributes": Object {
                  "label": "Doctor Octopus",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/10",
                "attributes": Object {
                  "label": "Elektra",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/11",
                "attributes": Object {
                  "label": "Black Widow",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/12",
                "attributes": Object {
                  "label": "Hawkeye",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/13",
                "attributes": Object {
                  "label": "Bullseye",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/14",
                "attributes": Object {
                  "label": "TaskMaster",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/15",
                "attributes": Object {
                  "label": "Star-Lord",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/16",
                "attributes": Object {
                  "label": "DareDevil",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/17",
                "attributes": Object {
                  "label": "Ikari",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/18",
                "attributes": Object {
                  "label": "War Machine",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/19",
                "attributes": Object {
                  "label": "Nick Fury",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/20",
                "attributes": Object {
                  "label": "Hank Pym",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/21",
                "attributes": Object {
                  "label": "Captain America",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
              Object {
                "$$key": "0/0/22",
                "attributes": Object {
                  "label": "Iron Man",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "0/0",
              },
            ],
          },
          "hasOptions": true,
          "parent": "0",
        },
        Object {
          "$$key": "0/1",
          "attributes": Object {
            "key": "test1",
            "label": "Mutants",
            "options": null,
          },
          "hasOptions": false,
          "parent": "0",
        },
        Object {
          "$$key": "0/2",
          "attributes": Object {
            "key": "test2",
            "label": "Celestials",
            "options": null,
          },
          "hasOptions": false,
          "parent": "0",
        },
      ],
    },
    "hasOptions": true,
    "parent": "root",
  },
  Object {
    "$$key": "1",
    "attributes": Object {
      "label": "DC Universe",
      "options": Array [
        Object {
          "$$key": "1/0",
          "attributes": Object {
            "label": "Humans",
            "options": null,
          },
          "hasOptions": false,
          "parent": "1",
        },
        Object {
          "$$key": "1/1",
          "attributes": Object {
            "label": "Meta Humans",
            "options": null,
          },
          "hasOptions": false,
          "parent": "1",
        },
        Object {
          "$$key": "1/2",
          "attributes": Object {
            "label": "The Quintessence",
            "options": Array [
              Object {
                "$$key": "1/2/0",
                "attributes": Object {
                  "label": "Zeus",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "1/2",
              },
              Object {
                "$$key": "1/2/1",
                "attributes": Object {
                  "label": "Ganthet",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "1/2",
              },
              Object {
                "$$key": "1/2/2",
                "attributes": Object {
                  "label": "Shazam",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "1/2",
              },
              Object {
                "$$key": "1/2/3",
                "attributes": Object {
                  "label": "Highfather",
                  "options": null,
                },
                "hasOptions": false,
                "parent": "1/2",
              },
            ],
          },
          "hasOptions": true,
          "parent": "1",
        },
      ],
    },
    "hasOptions": true,
    "parent": "root",
  },
  Object {
    "$$key": "2",
    "attributes": Object {
      "label": "Starcraft Universe",
      "options": Array [
        Object {
          "$$key": "2/0",
          "attributes": Object {
            "label": "Terrans",
            "options": null,
          },
          "hasOptions": false,
          "parent": "2",
        },
        Object {
          "$$key": "2/1",
          "attributes": Object {
            "label": "Zergs",
            "options": null,
          },
          "hasOptions": false,
          "parent": "2",
        },
        Object {
          "$$key": "2/2",
          "attributes": Object {
            "label": "Protos",
            "options": null,
          },
          "hasOptions": false,
          "parent": "2",
        },
      ],
    },
    "hasOptions": true,
    "parent": "root",
  },
  Object {
    "$$key": "3",
    "attributes": Object {
      "label": "Cartoon Network",
      "options": null,
    },
    "hasOptions": false,
    "parent": "root",
  },
  Object {
    "$$key": "4",
    "attributes": Object {
      "label": "Dragon Ball",
      "options": null,
    },
    "hasOptions": false,
    "parent": "root",
  },
]
`;

exports[`ListBoxBrowser should return option with $$key === 0 1`] = `
Object {
  "$$key": "0",
  "attributes": Object {
    "label": "Marvel Universe",
    "options": Array [
      Object {
        "$$key": "0/0",
        "attributes": Object {
          "label": "Humans",
          "options": Array [
            Object {
              "$$key": "0/0/0",
              "attributes": Object {
                "label": "Black Cat",
                "options": Array [
                  Object {
                    "$$key": "0/0/0/0",
                    "attributes": Object {
                      "label": "Akhenaten",
                      "options": null,
                    },
                    "hasOptions": false,
                    "parent": "0/0/0",
                  },
                  Object {
                    "$$key": "0/0/0/1",
                    "attributes": Object {
                      "label": "Alistair Smythe",
                      "options": null,
                    },
                    "hasOptions": false,
                    "parent": "0/0/0",
                  },
                  Object {
                    "$$key": "0/0/0/2",
                    "attributes": Object {
                      "label": "Angelo Fortunato",
                      "options": null,
                    },
                    "hasOptions": false,
                    "parent": "0/0/0",
                  },
                  Object {
                    "$$key": "0/0/0/3",
                    "attributes": Object {
                      "label": "Arcade",
                      "options": null,
                    },
                    "hasOptions": false,
                    "parent": "0/0/0",
                  },
                  Object {
                    "$$key": "0/0/0/4",
                    "attributes": Object {
                      "label": "Aria",
                      "options": null,
                    },
                    "hasOptions": false,
                    "parent": "0/0/0",
                  },
                  Object {
                    "$$key": "0/0/0/5",
                    "attributes": Object {
                      "label": "Baal",
                      "options": null,
                    },
                    "hasOptions": false,
                    "parent": "0/0/0",
                  },
                ],
              },
              "hasOptions": true,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/1",
              "attributes": Object {
                "label": "Silver Sable",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/2",
              "attributes": Object {
                "label": "Red Skull",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/3",
              "attributes": Object {
                "label": "Vulture",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/4",
              "attributes": Object {
                "label": "Falcon",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/5",
              "attributes": Object {
                "label": "Kraven the hunter",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/6",
              "attributes": Object {
                "label": "Bruce banner",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/7",
              "attributes": Object {
                "label": "Moon Knight",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/8",
              "attributes": Object {
                "label": "KingPin",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/9",
              "attributes": Object {
                "label": "Doctor Octopus",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/10",
              "attributes": Object {
                "label": "Elektra",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/11",
              "attributes": Object {
                "label": "Black Widow",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/12",
              "attributes": Object {
                "label": "Hawkeye",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/13",
              "attributes": Object {
                "label": "Bullseye",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/14",
              "attributes": Object {
                "label": "TaskMaster",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/15",
              "attributes": Object {
                "label": "Star-Lord",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/16",
              "attributes": Object {
                "label": "DareDevil",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/17",
              "attributes": Object {
                "label": "Ikari",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/18",
              "attributes": Object {
                "label": "War Machine",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/19",
              "attributes": Object {
                "label": "Nick Fury",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/20",
              "attributes": Object {
                "label": "Hank Pym",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/21",
              "attributes": Object {
                "label": "Captain America",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
            Object {
              "$$key": "0/0/22",
              "attributes": Object {
                "label": "Iron Man",
                "options": null,
              },
              "hasOptions": false,
              "parent": "0/0",
            },
          ],
        },
        "hasOptions": true,
        "parent": "0",
      },
      Object {
        "$$key": "0/1",
        "attributes": Object {
          "key": "test1",
          "label": "Mutants",
          "options": null,
        },
        "hasOptions": false,
        "parent": "0",
      },
      Object {
        "$$key": "0/2",
        "attributes": Object {
          "key": "test2",
          "label": "Celestials",
          "options": null,
        },
        "hasOptions": false,
        "parent": "0",
      },
    ],
  },
  "hasOptions": true,
  "parent": "root",
}
`;

exports[`ListBoxBrowser should return option with $$key === 1 1`] = `
Object {
  "$$key": "1",
  "attributes": Object {
    "label": "DC Universe",
    "options": Array [
      Object {
        "$$key": "1/0",
        "attributes": Object {
          "label": "Humans",
          "options": null,
        },
        "hasOptions": false,
        "parent": "1",
      },
      Object {
        "$$key": "1/1",
        "attributes": Object {
          "label": "Meta Humans",
          "options": null,
        },
        "hasOptions": false,
        "parent": "1",
      },
      Object {
        "$$key": "1/2",
        "attributes": Object {
          "label": "The Quintessence",
          "options": Array [
            Object {
              "$$key": "1/2/0",
              "attributes": Object {
                "label": "Zeus",
                "options": null,
              },
              "hasOptions": false,
              "parent": "1/2",
            },
            Object {
              "$$key": "1/2/1",
              "attributes": Object {
                "label": "Ganthet",
                "options": null,
              },
              "hasOptions": false,
              "parent": "1/2",
            },
            Object {
              "$$key": "1/2/2",
              "attributes": Object {
                "label": "Shazam",
                "options": null,
              },
              "hasOptions": false,
              "parent": "1/2",
            },
            Object {
              "$$key": "1/2/3",
              "attributes": Object {
                "label": "Highfather",
                "options": null,
              },
              "hasOptions": false,
              "parent": "1/2",
            },
          ],
        },
        "hasOptions": true,
        "parent": "1",
      },
    ],
  },
  "hasOptions": true,
  "parent": "root",
}
`;

exports[`ListBoxBrowser should return option with $$key === 1/2/1 1`] = `
Object {
  "$$key": "1/2/1",
  "attributes": Object {
    "label": "Ganthet",
    "options": null,
  },
  "hasOptions": false,
  "parent": "1/2",
}
`;

exports[`ListBoxBrowser should return option with $$key === 2/1 1`] = `
Object {
  "$$key": "2/1",
  "attributes": Object {
    "label": "Zergs",
    "options": null,
  },
  "hasOptions": false,
  "parent": "2",
}
`;
